File: BladeCraft/BladeCraft/Classes/Objects/Actions/actReactionBubble.cs
using System;
using System.Collections.Generic;
using System.Text;
using System.IO;

namespace BladeCraft.Classes.Objects.Actions
{
   public class actReactionBubble : Action
    {
        public string bubble, target;
        public bool loop, wait;
        public bool close;
        public float duration;

        public actReactionBubble(string bubble, string target, float duration, bool loop, bool wait)
        {
            this.bubble = bubble;
            this.duration = duration;
            this.target = target;
            this.loop = loop;
            this.wait = wait;
            this.close = false;
            actionType = type.ReactionBubble;
        }

        public actReactionBubble(string target)
        {
           this.target = target;
           this.close = true;
           actionType = type.ReactionBubble;
        }

        public actReactionBubble(actReactionBubble other)
        {
           this.bubble = other.bubble;
           this.duration = other.duration;
           this.target = other.target;
           this.loop = other.loop;
           this.wait = other.wait;
           this.close = other.close;
           actionType = type.ReactionBubble;
        }

        public override void write(StreamWriter writer) 
        {
           if (close)           
               writer.WriteLine(
                   "action closebubble " + target);
           else
              writer.WriteLine(
                   "action openbubble " + 
                   bubble + " " + target
                   + " " + duration.ToString() 
                   + " " + loop.ToString() 
                   + " " + wait.ToString());

        }
    }
}

File: BladeQuest/src/bladequest/actions/actReactionBubble.java
package bladequest.actions;

import bladequest.graphics.ReactionBubble;
import bladequest.world.GameObject;
import bladequest.world.Global;


public class actReactionBubble extends Action 
{
	private String target;
	private ReactionBubble bubble;
	private float duration;
	private boolean wait, loop, close;
	
	public actReactionBubble(String name, String target, float duration, boolean loop, boolean wait)
	{
		super();
		this.bubble = new ReactionBubble(Global.reactionBubbles.get(name));
		this.target = target;
		this.duration = duration;
		this.loop = loop;
		this.wait = wait;
		this.close = false;
	}
	
	public actReactionBubble(String target)
	{
		this.target = target;
		this.close = true;
	}
	
	@Override
	public void run()
	{
		
		if(target.equals("party"))
		{
			if(close)
				Global.party.closeReactionBubble();
			else
				Global.party.openReactionBubble(bubble, duration, loop);
		}
		else
		{
			for(GameObject go : Global.map.Objects())
			{
				if(go.Name().equals(target))
				{
					if(close)
						go.closeReactionBubble();
					else
						go.openReactionBubble(bubble, duration, loop);
					break;
				}
					
			}
		}
	}
	
	@Override
	public boolean isDone()
	{
		if(wait)
			return bubble.isDone();
		else
			return true;
	}

}
